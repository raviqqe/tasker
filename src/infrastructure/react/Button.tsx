import styled from "styled-components";
import { boxShadow } from "./style";
import { red, grey, white } from "./style/colors";

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? grey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
