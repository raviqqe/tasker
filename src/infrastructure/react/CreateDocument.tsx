import { MdAdd } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { InsertFilesFunction } from "./utilities";
import { MarkdownTextArea } from "./MarkdownTextArea";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({ createDocument, insertFiles }: IProps) => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
