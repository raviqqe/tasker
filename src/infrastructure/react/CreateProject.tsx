import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createProject: (name: string) => Promise<void>;
}

export const CreateProject = ({ createProject, ...restProps }: IProps) => (
  <CircleButton
    aria-label="Create Project"
    onClick={async () => {
      const name = window.prompt("Project name?");

      if (!name) {
        return;
      }

      await createProject(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
