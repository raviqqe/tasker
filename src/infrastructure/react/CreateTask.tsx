import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  createTask: (name: string) => Promise<void>;
}

export const CreateTask = ({ createTask }: IProps) => {
  return (
    <CircleButton
      aria-label="Create"
      onClick={async () => {
        const name = window.prompt("Task name?");

        if (!name) {
          return;
        }

        await createTask(name);
      }}
    >
      <MdAdd />
    </CircleButton>
  );
};
