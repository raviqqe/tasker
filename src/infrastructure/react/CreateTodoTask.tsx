import { MdAdd } from "react-icons/md";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createTodoTask: (name: string) => Promise<void>;
}

export const CreateTodoTask = ({ createTodoTask, ...restProps }: IProps): JSX.Element => (
  <CircleButton
    aria-label="Create"
    onClick={async () => {
      const name = window.prompt("Task name?");

      if (!name) {
        return;
      }

      await createTodoTask(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
