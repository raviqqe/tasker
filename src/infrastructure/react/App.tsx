import { useState } from "react";
import { useAsync } from "react-use";
import { styled } from "@linaria/react";
import { Home, type Props as HomeProps } from "./Home.js";
import { type Props as LandingProps, Landing } from "./Landing.js";
import { Loader } from "./Loader.js";
import { ProjectMenu, type Props as ProjectMenuProps } from "./ProjectMenu.js";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

export interface Props
  extends Omit<HomeProps, "showProjects">,
    Omit<ProjectMenuProps, "hideProjects">,
    LandingProps {
  initialize: () => Promise<void>;
  signedIn: boolean | null;
}

export const App = ({
  archivedProjects,
  archiveProject,
  createProject,
  currentProject,
  deleteProject,
  doneTasks,
  initialize,
  projects,
  repositoryUrl,
  signedIn,
  signIn,
  signOut,
  switchCurrentProject,
  todoTasks,
  unarchiveProject,
  updateProject,
  ...props
}: Props): JSX.Element => {
  useAsync(initialize, []);
  const [projectsShown, setProjectsShown] = useState(false);

  return signedIn === null ? (
    <LoaderContainer>
      <Loader />
    </LoaderContainer>
  ) : signedIn && projectsShown ? (
    <ProjectMenu
      archiveProject={archiveProject}
      archivedProjects={archivedProjects}
      createProject={createProject}
      currentProject={currentProject}
      deleteProject={deleteProject}
      hideProjects={() => setProjectsShown(false)}
      projects={projects}
      switchCurrentProject={switchCurrentProject}
      unarchiveProject={unarchiveProject}
      updateProject={updateProject}
    />
  ) : signedIn ? (
    <Home
      {...props}
      currentProject={currentProject}
      doneTasks={doneTasks}
      showProjects={() => setProjectsShown(true)}
      signOut={signOut}
      todoTasks={todoTasks}
    />
  ) : (
    <Landing repositoryUrl={repositoryUrl} signIn={signIn} />
  );
};
