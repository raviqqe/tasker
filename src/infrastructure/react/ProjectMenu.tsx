import { defaultImport } from "default-import";
import { useState, useRef, useEffect } from "react";
import { styled } from "@linaria/react";
import { type IProject } from "../../domain/project.js";
import {
  CreateProject,
  type IProps as ICreateProjectProps,
} from "./CreateProject.js";
import { Loader } from "./Loader.js";
import { Project, type IProps as IProjectProps } from "./Project.js";
import { ToggleProjects } from "./ToggleProjects.js";
import { grey, white, lightGrey } from "./style/colors.js";
import { boxShadow } from "./style.js";



const Container = styled.div`
  background-color: ${lightGrey};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const ScrollContainer = styled.div`
  overflow: auto;
`;

const ProjectsContainer = styled.div`
  ${boxShadow}
  display: flex;
  flex-direction: column;
  margin: 1.5rem;
  max-width: 50ex;
  background-color: ${white};
  border-radius: 0.5rem;
  padding: 1rem 1.5rem;

  > :not(:first-child) {
    margin-top: 0.75rem;
  }
`;

const Message = styled.div`
  color: ${grey};
`;

const LowerButtonsContainer = styled.div`
  position: fixed;
  right: 0.5rem;
  bottom: 0.5rem;

  > * {
    margin-top: 0.5rem;
  }
`;

const StyledCreateProject = styled(CreateProject)<{
  projectsArchived: boolean;
}>`
  visibility: ${({ projectsArchived }) =>
    projectsArchived ? "hidden" : "visible"};
`;

export interface IProps
  extends ICreateProjectProps,
    Required<Omit<IProjectProps, "currentProject" | "project">> {
  archivedProjects: IProject[] | null;
  currentProject: IProject | null;
  hideProjects: () => void;
  projects: IProject[] | null;
}

export const ProjectMenu = ({
  archivedProjects,
  archiveProject,
  createProject,
  currentProject,
  deleteProject,
  hideProjects,
  projects,
  switchCurrentProject,
  unarchiveProject,
  updateProject,
}: IProps): JSX.Element => {
  const [projectsArchived, setProjectsArchived] = useState(false);
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      ref.current.scrollIntoView({ block: "center" });
    }
  });

  return currentProject && archivedProjects && projects ? (
    <Container>
      <ScrollContainer>
        {projectsArchived ? (
          <ProjectsContainer>
            {archivedProjects.length === 0 ? (
              <Message>No archived projects</Message>
            ) : (
              archivedProjects.map((project) => (
                <Project
                  deleteProject={deleteProject}
                  key={project.id}
                  project={project}
                  unarchiveProject={async (project) => {
                    await unarchiveProject(project);
                    hideProjects();
                  }}
                />
              ))
            )}
          </ProjectsContainer>
        ) : (
          <ProjectsContainer>
            {projects.map((project) => (
              <Project
                archiveProject={archiveProject}
                currentProject={currentProject}
                key={project.id}
                project={project}
                ref={project.id === currentProject.id ? ref : null}
                switchCurrentProject={async (project) => {
                  hideProjects();
                  await switchCurrentProject(project);
                }}
                updateProject={updateProject}
              />
            ))}
          </ProjectsContainer>
        )}
      </ScrollContainer>
      <LowerButtonsContainer>
        <ToggleProjects
          projectsArchived={projectsArchived}
          setProjectsArchived={setProjectsArchived}
        />
        <StyledCreateProject
          createProject={async (name: string): Promise<void> => {
            await createProject(name);
            hideProjects();
          }}
          projectsArchived={projectsArchived}
        />
      </LowerButtonsContainer>
    </Container>
  ) : (
    <Container>
      <Loader />
    </Container>
  );
};
