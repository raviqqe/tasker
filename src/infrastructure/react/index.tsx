import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { IDocument } from "../../domain/document";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { GlobalStyle } from "./style";
import { App } from "./App";

export class ReactRenderer {
  constructor(
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly authenticationStore: AuthenticationStore,
    private readonly documentsStore: DocumentsStore,
    private readonly repositoryURL: string
  ) {}

  public render(element: HTMLElement): void {
    render(
      <>
        <App
          authenticationStore={this.authenticationStore}
          documentsStore={this.documentsStore}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument, text: string) =>
            this.documentUpdater.update(document, text)
          }
        />
        <GlobalStyle />
      </>,
      element
    );
  }
}
