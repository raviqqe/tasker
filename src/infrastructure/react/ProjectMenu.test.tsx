import { render } from "@testing-library/react";
import { atom } from "nanostores";
import { beforeEach, expect, it, vi } from "vitest";
import { projectPresenter } from "../../main/project-presenter.js";
import { ProjectMenu } from "./ProjectMenu.js";

beforeEach(() => {
  vi.spyOn(projectPresenter, "currentProject", "get").mockReturnValue(
    atom({ archived: false, id: "", name: "" }),
  );
});

it("renders", () => {
  expect(
    render(<ProjectMenu onHideProjects={() => {}} />).container.firstChild,
  ).toMatchSnapshot();
});
