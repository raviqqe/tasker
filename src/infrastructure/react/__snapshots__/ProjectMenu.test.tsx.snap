// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div
  class="sc-gKcmaw iINgjV"
>
  <div
    class="sc-iCfLYb bClTbH"
  >
    <div
      class="sc-furvoW koKzkQ"
    />
  </div>
  <div
    class="sc-jrQyNv hKbFyV"
  >
    <button
      aria-label="Toggle Projects"
      class="sc-bdvuGq sc-gsDLol jPrSNB enJXvF"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M20.55 5.22l-1.39-1.68A1.51 1.51 0 0018 3H6c-.47 0-.88.21-1.15.55L3.46 5.22C3.17 5.57 3 6.01 3 6.5V19a2 2 0 002 2h14c1.1 0 2-.9 2-2V6.5c0-.49-.17-.93-.45-1.28zM12 9.5l5.5 5.5H14v2h-4v-2H6.5L12 9.5zM5.12 5l.82-1h12l.93 1H5.12z"
        />
      </svg>
    </button>
    <button
      aria-label="Create Project"
      class="sc-bdvuGq sc-gsDLol nounu enJXvF sc-kDTjaY imDgJH"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
